File: presentation/src/main/java/info/firozansari/presentation/presenter/UserDetailsPresenter.java

package info.firozansari.presentation.presenter;

import android.support.annotation.NonNull;
import info.firozansari.domain.User;
import info.firozansari.domain.exception.DefaultErrorBundle;
import info.firozansari.domain.exception.ErrorBundle;
import info.firozansari.domain.interactor.DefaultObserver;
import info.firozansari.domain.interactor.GetUserDetails;
import info.firozansari.domain.interactor.GetUserDetails.Params;
import info.firozansari.presentation.exception.ErrorMessageFactory;
import info.firozansari.presentation.internal.di.PerActivity;
import info.firozansari.presentation.mapper.UserModelDataMapper;
import info.firozansari.presentation.model.UserModel;
import info.firozansari.presentation.view.UserDetailsView;
import javax.inject.Inject;

/**
 * {@link Presenter} that controls communication between views and models of the presentation
 * layer.
 */
@PerActivity
public class UserDetailsPresenter implements Presenter {

  private UserDetailsView viewDetailsView;

  private final GetUserDetails getUserDetailsUseCase;
  private final UserModelDataMapper userModelDataMapper;

  @Inject
  public UserDetailsPresenter(GetUserDetails getUserDetailsUseCase,
      UserModelDataMapper userModelDataMapper) {
    this.getUserDetailsUseCase = getUserDetailsUseCase;
    this.userModelDataMapper = userModelDataMapper;
  }

  public void setView(@NonNull UserDetailsView view) {
    this.viewDetailsView = view;
  }

  @Override public void resume() {}

  @Override public void pause() {}

  @Override public void destroy() {
    this.getUserDetailsUseCase.dispose();
    this.viewDetailsView = null;
  }

  /**
   * Initializes the presenter by showing/hiding proper views
   * and retrieving user details.
   */
  public void initialize(int userId) {
    this.hideViewRetry();
    this.showViewLoading();
    this.getUserDetails(userId);
  }

  private void getUserDetails(int userId) {
    this.getUserDetailsUseCase.execute(new UserDetailsObserver(), Params.forUser(userId));
  }

  private void showViewLoading() {
    this.viewDetailsView.showLoading();
  }

  private void hideViewLoading() {
    this.viewDetailsView.hideLoading();
  }

  private void showViewRetry() {
    this.viewDetailsView.showRetry();
  }

  private void hideViewRetry() {
    this.viewDetailsView.hideRetry();
  }

  private void showErrorMessage(ErrorBundle errorBundle) {
    String errorMessage = ErrorMessageFactory.create(this.viewDetailsView.context(),
        errorBundle.getException());
    this.viewDetailsView.showError(errorMessage);
  }

  private void showUserDetailsInView(User user) {
    final UserModel userModel = this.userModelDataMapper.transform(user);
    this.viewDetailsView.renderUser(userModel);
  }

  private final class UserDetailsObserver extends DefaultObserver<User> {

    @Override public void onComplete() {
      UserDetailsPresenter.this.hideViewLoading();
    }

    @Override public void onError(Throwable e) {
      UserDetailsPresenter.this.hideViewLoading();
      UserDetailsPresenter.this.showErrorMessage(new DefaultErrorBundle((Exception) e));
      UserDetailsPresenter.this.showViewRetry();
    }

    @Override public void onNext(User user) {
      UserDetailsPresenter.this.showUserDetailsInView(user);
    }
  }
}

File: presentation/src/main/java/info/firozansari/presentation/presenter/UserDetailsPresenter.kt
package info.firozansari.presentation.presenter

import info.firozansari.domain.User
import info.firozansari.domain.exception.DefaultErrorBundle
import info.firozansari.domain.exception.ErrorBundle
import info.firozansari.domain.interactor.DefaultObserver
import info.firozansari.domain.interactor.GetUserDetails
import info.firozansari.domain.interactor.GetUserDetails.Params.Companion.forUser
import info.firozansari.presentation.exception.ErrorMessageFactory
import info.firozansari.presentation.internal.di.PerActivity
import info.firozansari.presentation.mapper.UserModelDataMapper
import info.firozansari.presentation.view.UserDetailsView
import javax.inject.Inject

/**
 * [Presenter] that controls communication between views and models of the presentation
 * layer.
 */
@PerActivity
class UserDetailsPresenter @Inject constructor(
    private val getUserDetailsUseCase: GetUserDetails?,
    private val userModelDataMapper: UserModelDataMapper?
) : Presenter {
    private var viewDetailsView: UserDetailsView? = null
    fun setView(view: UserDetailsView) {
        viewDetailsView = view
    }

    override fun resume() {}
    override fun pause() {}
    override fun destroy() {
        getUserDetailsUseCase.dispose()
        viewDetailsView = null
    }

    /**
     * Initializes the presenter by showing/hiding proper views
     * and retrieving user details.
     */
    fun initialize(userId: Int) {
        hideViewRetry()
        showViewLoading()
        getUserDetails(userId)
    }

    private fun getUserDetails(userId: Int) {
        getUserDetailsUseCase.execute(UserDetailsObserver(), forUser(userId))
    }

    private fun showViewLoading() {
        viewDetailsView.showLoading()
    }

    private fun hideViewLoading() {
        viewDetailsView.hideLoading()
    }

    private fun showViewRetry() {
        viewDetailsView.showRetry()
    }

    private fun hideViewRetry() {
        viewDetailsView.hideRetry()
    }

    private fun showErrorMessage(errorBundle: ErrorBundle?) {
        val errorMessage = ErrorMessageFactory.create(
            viewDetailsView.context(),
            errorBundle.getException()
        )
        viewDetailsView.showError(errorMessage)
    }

    private fun showUserDetailsInView(user: User?) {
        val userModel = userModelDataMapper.transform(user)
        viewDetailsView.renderUser(userModel)
    }

    private inner class UserDetailsObserver : DefaultObserver<User?>() {
        override fun onComplete() {
            hideViewLoading()
        }

        override fun onError(e: Throwable?) {
            hideViewLoading()
            showErrorMessage(DefaultErrorBundle(e as Exception?))
            showViewRetry()
        }

        override fun onNext(user: User?) {
            showUserDetailsInView(user)
        }
    }
}